src/hotspot/share/opto/vectorization.hpp

```diff
@@ -667,340 +698,298 @@ class VLoopAnalyzer : StackObj {
   VStatus setup_submodules_helper();
 };
 
-// A vectorization pointer (VPointer) has information about an address for
-// dependence checking and vector alignment. It's usually bound to a memory
-// operation in a counted loop for vectorizable analysis.
-//
-// We parse and represent pointers of the simple form:
-//
-//   pointer   = adr + offset + invar + scale * ConvI2L(iv)
+// Reminder: MemPointer have the form:
 //
-// Where:
-//
-//   adr: the base address of an array (base = adr)
-//        OR
-//        some address to off-heap memory (base = TOP)
+//   pointer = SUM(summands) + con
 //
-//   offset: a constant offset
-//   invar:  a runtime variable, which is invariant during the loop
-//   scale:  scaling factor
-//   iv:     loop induction variable
+// Where every summand in summands has the form:
 //
-// But more precisely, we parse the composite-long-int form:
+//   summand = scale * variable
 //
-//   pointer   = adr + long_offset + long_invar + long_scale * ConvI2L(int_offset + inv_invar + int_scale * iv)
+// The VPointer wraps a MemPointer for the use in loops. A "valid" VPointer has
+// the form:
 //
-//   pointer   = adr + long_offset + long_invar + long_scale * ConvI2L(int_index)
-//   int_index =       int_offset  + int_invar  + int_scale  * iv
+//   pointer = base + invar + iv_scale * iv + con
 //
-// However, for aliasing and adjacency checks (e.g. VPointer::cmp()) we always use the simple form to make
-// decisions. Hence, we must make sure to only create a "valid" VPointer if the optimisations based on the
-// simple form produce the same result as the compound-long-int form would. Intuitively, this depends on
-// if the int_index overflows, but the precise conditions are given in VPointer::is_safe_to_use_as_simple_form().
+//   invar = SUM(invar_summands)
 //
-//   ConvI2L(int_index) = ConvI2L(int_offset  + int_invar  + int_scale  * iv)
-//                      = Convi2L(int_offset) + ConvI2L(int_invar) + ConvI2L(int_scale) * ConvI2L(iv)
+// Where:
+//   - base: is the known base of the MemPointer.
+//       on-heap (object base) or off-heap (native base address)
+//   - iv and iv_scale: i.e. the iv_summand = iv * iv_scale.
+//       If we find a summand where the variable is the iv, we set iv_scale to the
+//       corresponding scale. If there is no such summand, then we know that the
+//       pointer does not depend on the iv, since otherwise there would have to be
+//       a summand where its variable is main-loop variant.
+//   - invar_summands: all other summands except base and iv_summand.
+//       All variables must be pre-loop invariant. This is important when we need
+//       to memory align a pointer using the pre-loop limit.
 //
-//   scale  = long_scale * ConvI2L(int_scale)
-//   offset = long_offset + long_scale * ConvI2L(int_offset)
-//   invar  = long_invar  + long_scale * ConvI2L(int_invar)
+// A VPointer can be marked "invalid", if some of these conditions are not met, or
```

Comment:
You mention here "some of these conditions" which one could guess that you mean that it does not conform to this form. But further down you present the actual cases where a VPointer becomes invalid. Should we maybe first add these invalid cases and only then follow up with your explanation what this means for the aliasing queries? (just swap the two sections)

Reply:
Maybe add "provably" since you mention "or unknown" in the next sentence.
```suggestion
// A VPointer can be marked "invalid", if some of these conditions are provably met, or
```

Reply:
Ok, I swapped it around, and removed an unnecessary sentence.